File: libs/src/typescript/salmon-fs/fs/file/ifile.ts
/*
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { RandomAccessStream } from "../../../salmon-core/streams/random_access_stream.js";
import { Credentials } from './credentials.js';

/**
 * Interface that represents a real file. This class is used internally by the virtual disk to
 * import, store, and export the encrypted files.
 * Extend this to provide an interface to any file system, platform, or API ie: on disk, memory, network, or cloud.
 */
export interface IFile {
    /**
     * True if this file exists.
     *
     * @returns {Promise<boolean>}
     */
    exists(): Promise<boolean>;

    /**
     * Delete this file.
     *
     * @returns {Promise<boolean>}
     */
    delete(): Promise<boolean>;

    /**
     * Get a stream for reading the file.
     *
     * @returns {Promise<RandomAccessStream>} The stream
     * @throws FileNotFoundException
     */
    getInputStream(): Promise<RandomAccessStream>;

    /**
     * Get a stream for writing to the file.
     *
     * @returns {Promise<RandomAccessStream>} The stream
     * @throws FileNotFoundException
     */
    getOutputStream(): Promise<RandomAccessStream>;

    /**
     * Rename file.
     *
     * @param {string} newFilename The new filename
     * @returns {Promise<boolean>} True if success.
     * @throws FileNotFoundException
     */
    renameTo(newFilename: string): Promise<boolean>;

    /**
     * Get the length for the file.
     *
     * @returns {Promise<number>} The length.
     */
    getLength(): Promise<number>;

    /**
     * Get the count of files and subdirectories
     *
     * @returns {Promise<number>} The number of files and subdirectories
     */
    getChildrenCount(): Promise<number>;

    /**
     * Get the last modified date of the file.
     *
     * @returns {Promise<number>} The last date modified
     */
    getLastDateModified(): Promise<number>;

    /**
     * Get the display path.
     *
     * @returns {string} The display path
     */
    getDisplayPath(): string;

    /**
     * Get the path or any handle of the file on disk.
     *
     * @returns {any} The path
     */
    getPath(): any;

    /**
     * True if this is a file.
     *
     * @returns {Promise<boolean>} True if file.
     */
    isFile(): Promise<boolean>;

    /**
     * True if this is a directory.
     *
     * @returns {Promise<boolean>} True if directory
     */
    isDirectory(): Promise<boolean>;

    /**
     * Get all files and directories under this directory.
     *
     * @returns {Promise<IFile[]>} The list of files and subdirectories
     */
    listFiles(): Promise<IFile[]>;

    /**
     * Get the name of the file.
     *
     * @returns {string} The file name
     */
    getName(): string;

    /**
     * Create the directory with the name provided under this directory.
     *
     * @param {string} dirName Directory name.
     * @returns {Promise<IFile>} The newly created directory.
     */
    createDirectory(dirName: string): Promise<IFile>;

    /**
     * Get the parent directory of this file/directory.
     *
     * @returns {Promise<IFile | null>} The parent directory.
     */
    getParent(): Promise<IFile | null>;

    /**
     * Create an empty file with the provided name.
     *
     * @param {string} filename The name for the new file.
     * @returns {Promise<IFile>} The newly create file.
     * @throws IOException Thrown if there is an IO error.
     */
    createFile(filename: string): Promise<IFile>;

    /**
     * Move this file to another directory.
     *
     * @param {IFile} newDir           The target directory.
     * @param {MoveOptions} [options]          The options
     * @returns {Promise<IFile>} The file after the move. Use this instance for any subsequent file operations.
     */
    move(newDir: IFile, options?: MoveOptions): Promise<IFile>;

    /**
     * Copy this file to another directory.
     *
     * @param {IFile} newDir           The target directory.
     * @param {CopyOptions} [options]          The options
     * @returns {Promise<IFile | null>} The file after the copy. Use this instance for any subsequent file operations.
     * @throws IOException Thrown if there is an IO error.
     */
    copy(newDir: IFile, options?: CopyOptions): Promise<IFile | null>;

    /**
     * Get the file/directory matching the name provided under this directory.
     *
     * @param {string} filename The name of the file or directory to match.
     * @returns {Promise<IFile | null>} The file that was matched.
     */
    getChild(filename: string): Promise<IFile | null>;

    /**
     * Create a directory with the current filepath.
     *
     * @returns {Promise<boolean>}
     */
    mkdir(): Promise<boolean>;
	
	/**
     * Reset cached properties
     */
    reset(): any;

    /**
	 * Get the credentials
	 * @return The credentials
	 */
    getCredentials(): Credentials | null;
}

/**
 * Copy contents of a file to another file.
 *
 * @param {IFile} src              The source directory
 * @param {IFile} dest             The target directory
 * @param {CopyContentOptions} [options] The options 
 * @returns {Promise<boolean>} True if files are copied successfully
 * @throws IOException Thrown if there is an IO error.
 */
export async function copyFileContents(src: IFile, dest: IFile, options?: CopyContentsOptions): Promise<boolean> {
    let source: RandomAccessStream = await src.getInputStream();
    let target: RandomAccessStream = await dest.getOutputStream();
    try {
        await source.copyTo(target, 0, options?.onProgressChanged);
    } catch (ex) {
        await dest.delete();
        return false;
    } finally {
        await source.close();
        await target.close();
    }
    return true;
}

/**
 * Copy a directory recursively
 *
 * @param {IFile} src Source directory
 * @param {IFile} destDir Destination directory to copy into.
 * @param {RecursiveCopyOptions} [options] The options.
 * @throws IOException Thrown if there is an IO error.
 */
export async function copyRecursively(src: IFile, destDir: IFile, options?: RecursiveCopyOptions | null): Promise<void> {
    if(!options)
        options = new RecursiveCopyOptions();
    let newFilename: string = src.getName();
    let newFile: IFile | null;
    newFile = await destDir.getChild(newFilename);
    if (await src.isFile()) {
        if (newFile  && await newFile.exists()) {
            if (options.autoRename) {
                newFilename = await options.autoRename(src);
            } else {
                if (options.onFailed)
                    options.onFailed(src, new Error("Another file exists"));
                return;
            }
        }
        let copyOptions = new CopyOptions();
        copyOptions.newFilename = newFilename;
        copyOptions.onProgressChanged = (position, length) => {
            if (options.onProgressChanged) {
                options.onProgressChanged(src, position, length);
            }
        };
        await src.copy(destDir, copyOptions);
    } else if (await src.isDirectory()) {
        if (options.onProgressChanged)
            options.onProgressChanged(src, 0, 1);
        if (destDir.getDisplayPath().startsWith(src.getDisplayPath())) {
            if (options.onProgressChanged)
                options.onProgressChanged(src, 1, 1);
            return;
        }
        if (newFile  && await newFile.exists() && options.autoRename  && options.autoRenameFolders)
            newFile = await destDir.createDirectory(await options.autoRename(src));
        else if (newFile == null || !await newFile.exists())
            newFile = await destDir.createDirectory(newFilename);
        if (options.onProgressChanged)
            options.onProgressChanged(src, 1, 1);
        for (let child of await src.listFiles()) {
            if (newFile == null)
                throw new Error("Could not get new file");
            await copyRecursively(child, newFile, options);
        }
    }
}

/**
 * Move a directory recursively
 *
 * @param {IFile} file Source directory
 * @param {IFile} destDir Destination directory to move into.
 * @param {RecursiveMoveOptions} [options] The options 
 */
export async function moveRecursively(file: IFile, destDir: IFile, options?: RecursiveMoveOptions | null): Promise<void> {
    if(!options)
        options = new RecursiveMoveOptions();
    // target directory is the same
    let parent: IFile | null = await file.getParent();
    if (parent  && parent.getDisplayPath() == destDir.getDisplayPath()) {
        if (options.onProgressChanged) {
            options.onProgressChanged(file, 0, 1);
            options.onProgressChanged(file, 1, 1);
        }
        return;
    }

    let newFilename: string = file.getName();
    let newFile: IFile | null;
    newFile = await destDir.getChild(newFilename);
    if (await file.isFile()) {
        if (newFile  && await newFile.exists()) {
            if (newFile.getDisplayPath() == file.getDisplayPath())
                return;
            if (options.autoRename) {
                newFilename = await options.autoRename(file);
            } else {
                if (options.onFailed)
                    options.onFailed(file, new Error("Another file exists"));
                return;
            }
        }
        let moveOptions: MoveOptions = new MoveOptions();
        moveOptions.newFilename = newFilename;
        moveOptions.onProgressChanged = (position: number, length: number) => {
            if (options.onProgressChanged) {
                options.onProgressChanged(file, position, length);
            }
        };
        await file.move(destDir, moveOptions);
    } else if (await file.isDirectory()) {
        if (options.onProgressChanged)
            options.onProgressChanged(file, 0, 1);
        if (destDir.getDisplayPath().startsWith(file.getDisplayPath())) {
            if (options.onProgressChanged)
                options.onProgressChanged(file, 1, 1);
            return;
        }
        if ((newFile  && await newFile.exists() && options.autoRename  && options.autoRenameFolders)
            || newFile == null || !await newFile.exists()) {
            if (options.autoRename) {
                let moveOptions: MoveOptions = new MoveOptions();
                moveOptions.newFilename = await options.autoRename(file);
                newFile = await file.move(destDir, moveOptions);
            }
            return;
        }
        if (options.onProgressChanged)
            options.onProgressChanged(file, 1, 1);

        for (let child of await file.listFiles()) {
            if (newFile == null)
                throw new Error("Could not get new file");
            await moveRecursively(child, newFile, options);
        }

        if (!await file.delete()) {
            if (options.onFailed)
                options.onFailed(file, new Error("Could not delete source directory"));
        }
    }
}

/**
 * Delete a directory recursively
 * @param {RecursiveDeleteOptions} [options] The options
 */
export async function deleteRecursively(file: IFile, options?: RecursiveDeleteOptions | null): Promise<void> {
    if(!options)
        options = new RecursiveDeleteOptions();
    if (await file.isFile()) {
        if(options.onProgressChanged)
            options.onProgressChanged(file, 0, 1);
        if (!file.delete()) {
            if (options.onFailed)
                options.onFailed(file, new Error("Could not delete file"));
        }
        if(options.onProgressChanged)
            options.onProgressChanged(file, 1, 1);
    } else if (await file.isDirectory()) {
        for (let child of await file.listFiles()) {
            await deleteRecursively(child, options);
        }
        if (await !file.delete()) {
            if (options.onFailed)
                options.onFailed(file, new Error("Could not delete directory"));
        }
    }
}

/**
 * Get an auto generated copy of the name for a file.
 * @param {IFile} file The file
 * @returns {Promise<string>} The new file name
 */
export async function autoRenameFile(file: IFile): Promise<string> {
    return autoRename(file.getName());
};

/**
 * Get an auto generated copy of a file name
 *
 * @param {string} filename The current file name
 * @returns {string} The new file name
 */
export function autoRename(filename: string): string {
    let ext: string = getExtension(filename);
    let filenameNoExt: string;
    if (ext.length > 0)
        filenameNoExt = filename.substring(0, filename.length - ext.length - 1);
    else
        filenameNoExt = filename;
    let date: Date = new Date();
    let newFilename: string = filenameNoExt + " (" + date.getHours().toString().padStart(2, "0")
        + date.getHours().toString().padStart(2, "0") + date.getMinutes().toString().padStart(2, "0")
        + date.getSeconds().toString().padStart(2, "0") + date.getMilliseconds().toString().padStart(3, "0") + ")";
    if (ext.length > 0)
        newFilename += "." + ext;
    return newFilename;
}

/**
 * Get extension from file name
 * @param {string} fileName The file name
 * @returns {string} The extension
 */
export function getExtension(fileName: string): string {
    if (fileName == null)
        return "";
    let index: number = fileName.lastIndexOf(".");
    if (index >= 0) {
        return fileName.substring(index + 1);
    } else
        return "";
}

/**
 * File copy options
 */
export class CopyOptions {
    /**
     * Override filename
     */
    newFilename?: string | null = undefined;

    /**
     * Callback where progress changed
	 * (position: number, length: number) => void
     */
    onProgressChanged?: ((position: number, length: number) => void)  = undefined;
}

/**
 * File move options
 */
export class MoveOptions {
    /**
     * Override filename
     */
    newFilename?: string = undefined;

    /**
     * Callback where progress changed
	 * (position: number, length: number) => void
     */
    onProgressChanged?: ((position: number, length: number) => void) = undefined;
}

/**
 * Directory copy options (recursively)
 */
export class RecursiveCopyOptions {
    /**
     * Callback when file with same name exists
	 * (file: IFile) => Promise<string>
     */
    autoRename?: ((file: IFile) => Promise<string>) = undefined;

    /**
     * True to autorename folders
     */
    autoRenameFolders: boolean = false;

    /**
     * Callback when file changes
	 * (file: IFile, ex: Error) => void
     */
    onFailed?: ((file: IFile, ex: Error) => void) = undefined;

    /**
     * Callback where progress changed
	 * (file: IFile, position: number, length: number) => void
     */
    onProgressChanged?: ((file: IFile, position: number, length: number) => void) = undefined;
}

/**
 * Directory move options (recursively)
 */
export class RecursiveMoveOptions {
    /**
     * Callback when file with the same name exists
	 * (file: IFile) => Promise<string>
     */
    autoRename?: ((file: IFile) => Promise<string>) = undefined;

    /**
     * True to autorename folders
     */
    autoRenameFolders: boolean = false;

    /**
     * Callback when file failed
     */
    onFailed?: ((file: IFile, ex: Error) => void) = undefined;
    
    /**
     * Callback when progress changes
	 * (file: IFile, position: number, length: number) => void
     */
    onProgressChanged?: ((file: IFile, position: number, length: number) => void) = undefined;
}

/**
 * Directory move options (recursively)
 */
export class RecursiveDeleteOptions {
    /**
     * Callback when file failed
	 * (file: IFile, ex: Error) => void
     */
    onFailed?: ((file: IFile, ex: Error) => void) = undefined;

    /**
     * Callback when progress changed
	 * (file: IFile, position: number, length: number) => void
     */
    onProgressChanged?: ((file: IFile, position: number, length: number) => void) = undefined;
}

/**
 * Directory move options (recursively)
 */
export class CopyContentsOptions {
	/**
      * Callback when progress changed
	  * (position: number, length: number) => void
      */
    onProgressChanged?: ((position: number, length: number) => void) = undefined;
}

File: libs/src/python/salmon_fs/salmon_fs/fs/file/ifile.py
#!/usr/bin/env python3
"""!@brief Interface that represents a real file
"""

from __future__ import annotations

__license__ = """
MIT License

Copyright (c) 2021 Max Kas

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from abc import ABC
from datetime import datetime
from typing import Callable, Any
from typeguard import typechecked

from salmon_core.streams.random_access_stream import RandomAccessStream


@typechecked
class IFile(ABC):
    """!
    Interface that represents a real file. This class is used internally by the virtual disk to
    import, store, and export the encrypted files.
    Extend this to provide an interface to any file system, platform, or API ie: on disk, memory, network, or cloud.
    See: {@link PyFile}
    """

    def exists(self) -> bool:
        """!
        True if this file exists.
        
        @returns True if exists
        """
        pass

    def delete(self) -> bool:
        """!
        Delete this file.
        
        @returns True if file deleted
        """
        pass

    def get_input_stream(self) -> RandomAccessStream:
        """!
        Get a stream for reading the file.
        
        @returns The stream
        @exception FileNotFoundException:         """
        pass

    def get_output_stream(self) -> RandomAccessStream:
        """!
        Get a stream for writing to the file.
        
        @returns The stream
        @exception FileNotFoundException:         """
        pass

    def rename_to(self, new_filename: str) -> bool:
        """!
        Rename file.
        
        @param new_filename: The new filename
        @returns True if success.
        @exception FileNotFoundException:         """
        pass

    def get_length(self) -> int:
        """!
        Get the length for the file.
        
        @returns The length.
        """
        pass

    def get_children_count(self) -> int:
        """!
        Get the count of files and subdirectories
        
        @returns The children count
        """
        pass

    def get_last_date_modified(self) -> int:
        """!
        Get the last modified date of the file.
        
        @returns The last modified date in milliseconds
        """
        pass

    def get_display_path(self) -> str:
        """!
        Get the absolute path of the file on disk.
        
        @returns The absolute path
        """
        pass

    def get_path(self) -> str:
        """!
        Get the original filepath of this file. This might symlinks or merged folders. To get the absolute path
        use {@link #getAbsolutePath()}.
        
        @returns The path
        """
        pass

    def is_file(self) -> bool:
        """!
        True if this is a file.
        
        @returns True if file
        """
        pass

    def is_directory(self) -> bool:
        """!
        True if this is a directory.
        
        @returns True if directory
        """
        pass

    def list_files(self) -> list[IFile]:
        """!
        Get all files and directories under this directory.
        
        @returns The files
        """
        pass

    def get_name(self) -> str:
        """!
        Get the basename of the file.
        
        @returns The base name
        """
        pass

    def create_directory(self, dir_name: str) -> IFile:
        """!
        Create the directory with the name provided under this directory.
        
        @param dir_name: Directory name.
        @returns The newly created directory.
        """
        pass

    def get_parent(self) -> IFile:
        """!
        Get the parent directory of this file/directory.
        
        @returns The parent directory.
        """
        pass

    def create_file(self, filename: str) -> IFile:
        """!
        Create an empty file with the provided name.
        
        @param filename: The name for the new file.
        @returns The newly create file.
        @exception IOError: Thrown if there is an IO error.
        """
        pass

    def move(self, new_dir: IFile, options: IFile.MoveOptions | None = None) -> IFile:
        """!
        Move this file to another directory.
        
        @param new_dir:           The target directory.
        @param options: The options
        @returns The file after the move. Use this instance for any subsequent file operations.
        """
        pass

    def copy(self, new_dir: IFile, options: IFile.CopyOptions | None = None) -> IFile:
        """!
        Copy this file to another directory.
        
        @param new_dir:           The target directory.
        @param options: The options
        @returns The file after the copy. Use this instance for any subsequent file operations.
        @exception IOError: Thrown if there is an IO error.
        """
        pass

    def get_child(self, filename: str) -> IFile:
        """!
        Get the file/directory matching the name provided under this directory.
        
        @param filename: The name of the file or directory to match.
        @returns The file that was matched.
        """
        pass

    def mkdir(self) -> bool:
        """!
        Create a directory with the current filepath.
        
        @returns True if directory created
        """
        pass

    def reset(self):
        """!
        Clear cached properties
        """
        pass

    def get_credentials(self):
        """!
        Get the credentials if available
        """
        pass

    @staticmethod
    def copy_file_contents(src: IFile, dest: IFile, options: IFile.CopyContentsOptions) -> bool:
        """!
        Copy contents of a file to another file.

        @param src:              The source directory
        @param dest:             The target directory
        @param options:     The options
        @returns True if contents copied
        @exception IOError: Thrown if there is an IO error.
        """
        source: RandomAccessStream = src.get_input_stream()
        target: RandomAccessStream = dest.get_output_stream()
        try:
            source.copy_to(target, on_progress_changed=options.on_progress_changed)
        except Exception as ex:
            dest.delete()
            return False
        finally:
            source.close()
            target.close()
        return True

    def copy_recursively(self, dest: IFile, options: IFile.RecursiveCopyOptions | None = None):
        """!
        Copy a directory recursively

        @param dest: The destination directory
        @param options: The options
        @exception IOError: Thrown if there is an IO error.
        """

        if not options:
            options = IFile.RecursiveCopyOptions()
        new_filename: str = self.get_name()
        new_file: IFile
        new_file = dest.get_child(new_filename)
        if self.is_file():
            if new_file and new_file.exists():
                if options.auto_rename:
                    new_filename = options.auto_rename(self)
                else:
                    if options.on_failed:
                        options.on_failed(self, Exception("Another directory/file exists"))
                    return
            copy_options = IFile.CopyOptions()
            copy_options.new_filename = new_filename
            copy_options.on_progress_changed = lambda position, length: self.notify(position, length,
                                                                                    options.on_progress_changed)
            self.copy(dest, copy_options)
        elif self.is_directory():
            if options.on_progress_changed:
                options.on_progress_changed(self, 0, 1)
            if dest.get_display_path().startswith(self.get_display_path()):
                if options.on_progress_changed:
                    options.on_progress_changed(self, 1, 1)
                return
            
            if new_file and new_file.exists() and options.auto_rename and options.auto_rename_folders:
                new_file = dest.create_directory(options.auto_rename(self))
            elif new_file is None or not new_file.exists():
                new_file = dest.create_directory(new_filename)
            elif new_file and new_file.exists() and new_file.is_file():
                if options.on_failed:
                    options.on_failed(self, Exception("Another file exists"))
                return

            if options.on_progress_changed:
                options.on_progress_changed(self, 1, 1)

            for child in self.list_files():
                child.copy_recursively(new_file, options)

    def move_recursively(self, dest: IFile, options: IFile.RecursiveMoveOptions | None = None):
        """!
        Move a directory recursively

        @param dest:              The target directory
        @param options: The options
        """
        if not options:
            options = IFile.RecursiveMoveOptions()

        # target directory is the same
        if self.get_parent().get_path() == dest.get_path():
            if options.on_progress_changed:
                options.on_progress_changed(self, 0, 1)
                options.on_progress_changed(self, 1, 1)
            return

        new_filename: str = self.get_name()
        new_file: IFile
        new_file = dest.get_child(new_filename)
        if self.is_file():
            if new_file and new_file.exists():
                if new_file.get_path() == self.get_path():
                    return
                if options.auto_rename:
                    new_filename = options.auto_rename(self)
                else:
                    if options.on_failed:
                        options.on_failed(self, Exception("Another directory/file exists"))
                    return

            move_options: IFile.MoveOptions = IFile.MoveOptions()
            move_options.new_filename = new_filename
            move_options.on_progress_changed = lambda position, length: self.notify(position, length,
                                                                                    options.on_progress_changed)
            self.move(dest, move_options)
        elif self.is_directory():
            if options.on_progress_changed:
                options.on_progress_changed(self, 0, 1)
            if dest.get_display_path().startswith(self.get_display_path()):
                if options.on_progress_changed:
                    options.on_progress_changed(self, 1, 1)
                return

            if new_file and new_file.exists() and options.auto_rename and options.auto_rename_folders:
                new_file = dest.create_directory(options.auto_rename(self))
            elif new_file is None or not new_file.exists():
                new_file = dest.create_directory(new_filename)
            elif new_file and new_file.exists() and new_file.is_file():
                if options.on_failed:
                    options.on_failed(self, Exception("Another file exists"))
                return

            if options.on_progress_changed:
                options.on_progress_changed(self, 1, 1)

            for child in self.list_files():
                child.move_recursively(new_file, options)

            if not self.delete():
                options.on_failed(self, Exception("Could not delete source directory"))

    def delete_recursively(self, options: IFile.RecursiveDeleteOptions | None = None):
        """!
        Delete a directory recursively
        @param options: The options
        """
        if not options:
            options = IFile.RecursiveDeleteOptions()

        if self.is_file():
            options.on_progress_changed(self, 0, 1)
            if not self.delete():
                options.on_failed(self, Exception("Could not delete file"))
                return

            options.on_progress_changed(self, 1, 1)
        elif self.is_directory():
            for child in self.list_files():
                child.delete_recursively(options)

            if not self.delete():
                if options.on_failed:
                    options.on_failed(self, Exception("Could not delete directory"))
                return

    @staticmethod
    def auto_rename_file(file: IFile) -> str:
        """!
        Get an auto generated copy of the name for a file.
        """
        return IFile.auto_rename(file.get_name())

    @staticmethod
    def auto_rename(filename: str) -> str:
        """!
        Get an auto generated copy of a filename

        @param filename:
        @returns The new file name
        """
        ext: str = IFile.__get_extension(filename)
        filename_no_ext: str | None = None
        if len(ext) > 0:
            filename_no_ext = filename[0:len(filename) - len(ext) - 1]
        else:
            filename_no_ext = filename
        new_filename: str = filename_no_ext + " (" + datetime.today().strftime("%H%m%S%f")[:-3] + ")"
        if len(ext) > 0:
            new_filename += "." + ext
        return new_filename

    def notify(self, position: int, length: int, progress_listener: Callable[[IFile, int, int], Any] | None = None):
        if progress_listener:
            progress_listener(self, position, length)

    @staticmethod
    def __get_extension(file_name: str) -> str:
        if file_name is None:
            return ""
        index: int = file_name.rindex(".") if "." in file_name else -1
        if index >= 0:
            return file_name[index + 1:]
        else:
            return ""

    @typechecked
    class CopyOptions:
        """!
        File copy options
        """

        new_filename: str | None = None
        """!
        Override filename
        """

        on_progress_changed: Callable[[int, int], Any] | None = None
        """!
        Callback where progress changed
        """

    @typechecked
    class MoveOptions:
        """!
        File move options
        """

        new_filename: str | None = None
        """!
        Override filename
        """

        on_progress_changed: Callable[[int, int], Any] | None = None
        """!
        Callback where progress changed
        """

    @typechecked
    class RecursiveCopyOptions:
        """!
        Directory copy options (recursively)
        """

        auto_rename: Callable[[IFile], str] | None = None
        """!
        Callback when file with same name exists
        """

        auto_rename_folders: bool = False
        """
        True to auto_rename folders
        """

        on_failed: Callable[[IFile, Exception], Any] | None = None
        """!
        Callback when file changes
        """

        on_progress_changed: Callable[[IFile, int, int], Any] | None = None
        """!
        Callback where progress changed
        """

    @typechecked
    class RecursiveMoveOptions:
        """!
        Directory move options (recursively)
        """

        auto_rename: Callable[[IFile], str] | None = None
        """!
        Callback when file with the same name exists
        """

        auto_rename_folders: bool = False
        """
        True to auto_rename folders
        """

        on_failed: Callable[[IFile, Exception], Any] | None = None
        """!
        Callback when file failed
        """

        on_progress_changed: Callable[[IFile, int, int], Any] | None = None
        """!
        Callback when progress changes
        """

    @typechecked
    class RecursiveDeleteOptions:
        """!
        Directory move options (recursively)
        """

        on_failed: Callable[[IFile, Exception], Any] | None = None
        """!
        Callback when file failed
        """

        on_progress_changed: Callable[[IFile, int, int], Any] | None = None
        """!
        Callback when progress changed
        """

    @typechecked
    class CopyContentsOptions:
        """!
        Directory move options (recursively)
        """

        on_progress_changed: Callable[[int, int], Any] | None = None
        """!
        Callback when progress changed
        """
